src/main/java/org/zalando/riptide/DefaultRoutingTree.java

```diff
@@ -0,0 +1,122 @@
+package org.zalando.riptide;
+
+/*
+ * ⁣​
+ * Riptide
+ * ⁣⁣
+ * Copyright (C) 2015 - 2016 Zalando SE
+ * ⁣⁣
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ​⁣
+ */
+
+import lombok.SneakyThrows;
+import org.springframework.http.client.ClientHttpResponse;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.Collections.unmodifiableMap;
+import static java.util.stream.Collectors.toMap;
+
+final class DefaultRoutingTree<A> implements RoutingTree<A> {
+
+    private final Navigator<A> navigator;
+    private final Map<A, Route> routes;
+    private final Optional<Route> wildcard;
+
+    DefaultRoutingTree(final Navigator<A> navigator, final List<Binding<A>> bindings) {
+        this(navigator, map(bindings));
+    }
+
+    private DefaultRoutingTree(final Navigator<A> navigator, final Map<A, Route> routes) {
+        this(navigator, unmodifiableMap(routes), Optional.ofNullable(routes.remove(null)));
```

Comment:
Works, but looks a bit strange to first wrap into an unmodifiable collection and then remove items from the backing collection.